### ISSUE TITLE
Passing  `as` prop while using styled notation creates hydration errors in NextJS

### ISSUE THREAD

**Issue by burningyouth** [styled-components/xstyled#404]
## 🐛 Bug Report

When `as` is passed to a component that uses basic styled-components notation in the NextJS app it creates hydration errors. 

Styles:
```jsx
const ButtonStyled = styled.buttonBox`
  border-radius: md;
  font-weight: semibold;
  transition: default;
  background-color: emerald-500;
  color: #fff;

  @media (min-width: md) {
    font-size: lg;
  }
`;

// This is strange code but let's pretend that we need some additional
// logic here and styled component is not enough
// Somewhat using ButtonStyled directly doesn't trigger error
export const Button = ({ children, ...props }: any) => {
  return <ButtonStyled {...props}>{children}</ButtonStyled>;
};
```

Using as:
```jsx 
<Button as="a" href="http://google.com">
          123
</Button>
```

Result: 
<img width="899" alt="Снимок экрана 2023-07-18 в 10 50 55" src="https://github.com/styled-components/xstyled/assets/21275316/c00d8b89-8b71-4bd5-905d-9a7af3fa5287">

## To Reproduce

Using the standard setup guide for styled-components in NextJS, create buttonBox component and use it in other component, pass `as="a"` and `href`, then look at the console. 

## Expected behavior

Using styled notation doesn't create errors.

## Link to repo

https://github.com/burningyouth/sc-nextjs-repr

```bash
## System:
 - OS: macOS 13.3
 - CPU: (8) arm64 Apple M1 Pro
 - Memory: 420.38 MB / 16.00 GB
 - Shell: 5.9 - /bin/zsh
## Binaries:
 - Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
 - Yarn: 1.22.15 - /usr/local/bin/yarn
 - npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
 - pnpm: 6.11.0 - /usr/local/bin/pnpm
```

**Comment by danilovIT** [comment]
Yes, I have the same problem

**Comment by quantizor** [comment]
This thread is from 2023; a lot has moved since then (xstyled 3.8+, styled-components v6, Next.js compiler changes, App Router). If you (or @danilovIT) can still reproduce on the current versions, an updated repro repo would be very helpful. Otherwise I'll close this in a couple weeks.
